### PR TITLE
Client Spec Test Rewrite

### PR DESCRIPTION
**Pull request by sighphyre**
### What is this?

This is a full rewrite of the current client specification tests to use the client specifications project, rather than a custom implementation of each test. This should load each specification test individually from file and execute them independently, while retaining the test name included in the spec file, so that when failures are picked up, we can know what's broken and why. This leverages the bootstrapping capability of this SDK to ensure data contracts are correct, but ignores the API level. Our plan to ensure API level correctness through an upcoming project across all of our officially supported SDKs.

### Why was this done?

There's a few reasons why this is a useful thing to do:
- Adding new test cases when Unleash adds new features is time consuming and error prone, we're planning on adding further tests in the near future to extend test case coverage to ensure that each of the official SDKs correctly support the edge cases and new features, I don't believe that anyone has appetite to replicate this work in the Python client
- On the Unleash side we're building an integration testing tool to validate all our SDKs against server API responses, while the approach in this project is not blocking to that, it has picked up a number of edge case failures
- Previously, this was tough to do, but with the addition of bootstrapping and the ability to set the client to read only mode this becomes much easier to tackle without mocking out the API responses

### Why is this a draft?

We've picked up a few test failures here, the plan is to dig a little deeper into those failures in the near future and either patch our specs if they're false negatives or fix the underlying issues

### CODE REVIEW

**Review by chriswk — COMMENTED** [review]
I feel like this is a good idea. Though, how does a test failure look with this new setup, is it still easy to see which tests are failing?

**Comment by sighphyre** [comment]
> I feel like this is a good idea. Though, how does a test failure look with this new setup, is it still easy to see which tests are failing?

This is a good question! This was important to to me because a few of our SDKs are badly behaved around reporting failures.  Taken right from my console:

```
FAILED tests/specification_tests/test_client_specs.py::test_spec[09-strategy-constraints-Feature.constraints.multi should be enabled in prod for user 2] - AssertionError: assert False == True
FAILED tests/specification_tests/test_client_specs.py::test_spec[09-strategy-constraints-Feature.constraints.custom should be enabled in prod for norway] - AssertionError: assert False == True
FAILED tests/specification_tests/test_client_specs.py::test_spec[09-strategy-constraints-Feature.constraints.dual should be enabled in prod] - AssertionError: assert False == True
FAILED tests/specification_tests/test_client_specs.py::test_spec[09-strategy-constraints-Enabled in prod for userId=123] - AssertionError: assert False == True
FAILED tests/specification_tests/test_client_specs.py::test_spec[12-custom-stickiness-Feature.flexible.rollout.custom.stickiness_100 should be enabled without field defined for 100%] - AssertionError: assert ...
FAILED tests/specification_tests/test_client_specs.py::test_spec[12-custom-stickiness-Feature.flexible.rollout.custom.stickiness_50 should be enabled for customField=388] - AssertionError: assert False == True
FAILED tests/specification_tests/test_client_specs.py::test_spec[12-custom-stickiness-Feature.flexible.rollout.custom.stickiness_50 should be enabled for customField=390] - AssertionError: assert False == True
FAILED tests/specification_tests/test_client_specs.py::test_spec[12-custom-stickiness-Feature.flexible.rollout.custom.stickiness_50 should be enabled for customField=391] - AssertionError: assert {'enabled': ...
FAILED tests/specification_tests/test_client_specs.py::test_spec[12-custom-stickiness-Feature.flexible.rollout.custom.stickiness_50 should be enabled for customField=392] - AssertionError: assert {'enabled': ...
FAILED tests/specification_tests/test_client_specs.py::test_spec[12-custom-stickiness-Feature.flexible.rollout.custom.stickiness_50 should be enabled for customField=393] - AssertionError: assert {'enabled': ...
FAILED tests/specification_tests/test_client_specs.py::test_spec[12-custom-stickiness-Feature.flexible.rollout.custom.stickiness_50 should be enabled for customField=394] - AssertionError: assert {'enabled': ...
FAILED tests/specification_tests/test_client_specs.py::test_spec[12-custom-stickiness-Feature.flexible.rollout.custom.stickiness_50 should be enabled for customField=395] - AssertionError: assert {'enabled': ...
FAILED tests/specification_tests/test_client_specs.py::test_spec[13-constraint-operators-F5.numEq] - AssertionError: assert False == True
FAILED tests/specification_tests/test_client_specs.py::test_spec[13-constraint-operators-F5.numEq works for floats] - AssertionError: assert False == True
```

So fairly similar to the Node SDK. Pytest will also report the exception but it's not all that useful:

```
________________________________________________________________________ test_spec[13-constraint-operators-F7.dateAfter should be enabled] ________________________________________________________________________

spec = (<UnleashClient.UnleashClient object at 0x7fb3281d6020>, {'context': {'appName': 'pytest', 'currentTime': '2022-01-30T...fault'}, 'description': 'F7.dateAfter should be enabled', 'expectedResult': True, 'toggleName': 'F7.dateAfter'}, False)

    @pytest.mark.parametrize("spec", TEST_DATA, ids=TEST_NAMES)
    def test_spec(spec):
        unleash_client, test_data, is_variant_test = spec
        if not is_variant_test:
            toggle_name = test_data["toggleName"]
            expected = test_data["expectedResult"]
            context = test_data.get("context")
>           assert unleash_client.is_enabled(toggle_name, context) == expected
E           AssertionError: assert False == True
E            +  where False = <bound method UnleashClient.is_enabled of <UnleashClient.UnleashClient object at 0x7fb3281d6020>>('F7.dateAfter', {'appName': 'pytest', 'currentTime': '2022-01-30T13:00:00.000Z', 'environment': 'default'})
E            +    where <bound method UnleashClient.is_enabled of <UnleashClient.UnleashClient object at 0x7fb3281d6020>> = <UnleashClient.UnleashClient object at 0x7fb3281d6020>.is_enabled

```

**Comment by chriswk** [comment]
That seems like a good ux though. Then it's on the spec writers to keep the cases clear.

File: UnleashClient/constraints/Constraint.py [hunk 1]
```diff
@@ -82,37 +82,39 @@ def check_string_operators(self, context_value: str) -> bool:
...
+            parsed_date = parse(self.value)
+            context_date = parse(context_value)
         except ParserError:
             LOGGER.error(f"Unable to parse date: {self.value}")
             parsing_exception = True
 
         if not parsing_exception:
             if self.operator == ConstraintOperators.DATE_AFTER:
-                return_value = context_value > parsed_date
+                return_value = context_date > parsed_date
```
[the remark this answers is not in the document]
Comment (sighphyre) [reply UnleashClient/constraints/Constraint.py:115]: Shhh sonatype. This is unreachable

File: UnleashClient/__init__.py [hunk 2]
```diff
@@ -279,8 +279,10 @@ def is_enabled(self,
         """
         context = context or {}
 
-        # Update context with static values
-        context.update(self.unleash_static_context)
+        base_context = self.unleash_static_context.copy()
+        # Update context with static values and allow context to override environment
+        base_context.update(context)
+        context = base_context
```
Comment (sighphyre) [inline UnleashClient/__init__.py:285]: Fixes an issue where the base context would overwrite the token's environment. This probably isn't a major issue in practice but it needs to be solved to get the client specs working fast

File: UnleashClient/constraints/Constraint.py [hunk 3]
```diff
@@ -82,37 +82,42 @@ def check_string_operators(self, context_value: str) -> bool:
...
+            return_value = parsed_context < parsed_value
         elif self.operator == ConstraintOperators.NUM_LTE:
-            return_value = context_value <= parsed_value
-
+            return_value = parsed_context <= parsed_value
         return return_value
 
 
-    def check_date_operators(self, context_value: datetime) -> bool:
+    def check_date_operators(self, context_value: Union[datetime, str]) -> bool:
```
Comment (sighphyre) [inline UnleashClient/constraints/Constraint.py:102]: I realise this is a very debatable change, however, all the other SDKs ingest a string at this point and that's also what our docs say. I've let a datetime continue to work here because there's a reasonable chance this is already in the wild

File: UnleashClient/strategies/FlexibleRolloutStrategy.py [hunk 4]
```diff
@@ -29,6 +29,7 @@ def apply(self, context: dict = None) -> bool:
         elif stickiness == "random":
             calculated_percentage = self.random_hash()
         else:
-            calculated_percentage = normalized_hash(context[stickiness], activation_group)
+            custom_stickiness = context.get(stickiness) or context.get("properties")[stickiness]
```
Comment (sighphyre) [inline UnleashClient/strategies/FlexibleRolloutStrategy.py:32]: Fixes an issue with stickiness where you have nested properties. I've allowed the exception raised by the null access here to propagate here when stickiness is not found on the dict rather than allow it to crash later

File: tests/unit_tests/test_constraints.py [hunk 5]
```diff
@@ -153,17 +153,17 @@ def test_constraints_NUM_FLOAT():
 def test_constraints_DATE_AFTER():
     constraint = Constraint(constraint_dict=mock_constraints.CONSTRAINT_DATE_AFTER)
 
-    assert constraint.apply({'currentTime': datetime(2022, 1, 23)})
-    assert not constraint.apply({'currentTime': datetime(2022, 1, 22)})
-    assert not constraint.apply({'currentTime': datetime(2022, 1, 21)})
+    assert constraint.apply({'currentTime': datetime(2022, 1, 23, tzinfo=pytz.UTC)})
```
Comment (sighphyre) [inline tests/unit_tests/test_constraints.py:156]: Patching tests to deal with the behaviour change - this SDK was incorrectly ignoring timezones

File: UnleashClient/constraints/Constraint.py (hunk 3 above)
Comment (ivanklee86) [reply UnleashClient/constraints/Constraint.py:102]: Oh that's interesting!  I was assuming we'd want to favor language-specific date constructs. 😓

**Review by ivanklee86 — CHANGES_REQUESTED** [review]
🎉 Looks awesome over!  Removes a lot of duplicated code, makes things easier to maintain.  c/ping new spec tests was the worst part of a feature.

I did have some questions about usability (since I'm a platform engineer by day)!

File: tests/specification_tests/test_client_specs.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
+import json
+import uuid
+from os import path
+from UnleashClient import UnleashClient
+from tests.utilities.testing_constants import URL, APP_NAME
+from UnleashClient.cache import FileCache
+import pytest
+
+
+CLIENT_SPEC_PATH = "tests/specification_tests/client-specification/specifications"
```
[the remark this answers is not in the document]
Comment (chriswk) [reply tests/specification_tests/test_client_specs.py:10]: The way we do it for the Java client is that we're explicitly targeting a version of the client-specification, so if new specifications are added that are not supported yet, they will fail if we try to target a newer client-specification until we've actually implemented it, but since you're locked to a specific version of the client-specifications, there shouldn't be a way a new case/spec can be added. From the GH workflow here, the python SDK does the same, it points to an specific version.

**Review by chriswk — APPROVED** [review]
I like it. Particularly the quick test run. :)

File: UnleashClient/constraints/Constraint.py [hunk 7]
```diff
@@ -82,37 +82,42 @@ def check_string_operators(self, context_value: str) -> bool:
...
+            else:
+                context_date = context_value
         except ParserError:
             LOGGER.error(f"Unable to parse date: {self.value}")
             parsing_exception = True
 
         if not parsing_exception:
             if self.operator == ConstraintOperators.DATE_AFTER:
-                return_value = context_value > parsed_date
+                return_value = context_date > parsed_date
```
[the remark this answers is not in the document]
Comment (sighphyre) [reply UnleashClient/constraints/Constraint.py:118]: @sonatype-lift ignore

File: .github/workflows/pull_request.yml [hunk 8]
```diff
@@ -59,6 +61,8 @@ jobs:
         run:  |
           python -m pip install --upgrade pip
           pip install tox
+      - name: Download test cases
+        run: git clone --depth 5 --branch v4.2.2 https://github.com/Unleash/client-specification.git tests/specification_tests/client-specification
```
[the remark this answers is not in the document]
Comment (sighphyre) [reply .github/workflows/pull_request.yml]: Yeah! That's a good plan, I've added it to the tox file and removed it from the tox pipeline

File: UnleashClient/constraints/Constraint.py (hunk 3 above)
Comment (sighphyre) [reply UnleashClient/constraints/Constraint.py:102]: Most of the client SDKs have grown up around the spec so they're a little more liberal because it can be a pain to guess the types in the context properties in client specs. I think the Python SDK made the right choices here, which is part of the reason I left this as a Union type

File: tests/specification_tests/test_client_specs.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sighphyre) [reply tests/specification_tests/test_client_specs.py:10]: What Chris said! The clone will pull from a tag in the repo so this is effectively pinning this to a tag in the client spec, which should never change. This does mean we'll have to remember to keep the pipeline up to date when we add new features but we're planning on handling monitoring feature completeness against the specs in a separate project

File: get-spec.sh [hunk 9]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+specversion=`python -c "from UnleashClient.constants import CLIENT_SPEC_VERSION; print(CLIENT_SPEC_VERSION)"`
```
[the remark this answers is not in the document]
Comment (sighphyre) [reply get-spec.sh:2]: @sonatype-lift ignore

Fixed

File: get-spec.sh [hunk 10]
```diff
@@ -0,0 +1,6 @@
...
+
+rm -rf tests/specification_tests/client-specification
+echo "Downloading client spec ${specversion}"
+git clone --depth 5 --branch v${specversion} https://github.com/Unleash/client-specification.git tests/specification_tests/client-specification
```
[the remark this answers is not in the document]
Comment (sighphyre) [reply get-spec.sh:6]: @sonatype-lift ignore

Fixed

**Review by ivanklee86 — APPROVED** [review]
